modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -344,44 +409,69 @@
+                  ;; there is no equivalent of .setMaxRows on a JDBC Statement; we rely on our middleware to stop
+                  ;; realizing more rows as per the maximum result size
+                  (.setMaxResults *page-size*))
+        ;; as long as we don't set certain additional QueryJobConfiguration options, our queries *should* always be
+        ;; following the fast query path (i.e. RPC)
+        ;; check out com.google.cloud.bigquery.QueryRequestInfo.isFastQuerySupported for full details
+        done-chan  (a/promise-chan)
+        res-fut    (future
+                     (log/trace "BigQuery exec thread sending query job")
+                     (try
+                       (let [result (.query client (.build request) (u/varargs BigQuery$JobOption))]
+                         (log/trace "BigQuery request finished successfully; sending the query to the channel")
+                         (let [pushed? (a/>!! done-chan [:done result])]
+                           (log/tracef "BigQuery channel send complete: %s" (pr-str pushed?)))
+                         nil)
+                       (catch Throwable t
+                         (a/>!! done-chan [:error t]))))]
+    (loop [[res chan] (a/alts!! (cond-> [done-chan]
+                                  cancel-chan (conj cancel-chan))
+                                :priority true)]
+      (if-let [[tag value] (when (= chan done-chan)
+                             res)]
+        ;; done-chan fired first, with either an :error or :done
+        (case tag
+          :done  value
+          :error (condp instance? value
+                   java.util.concurrent.CancellationException
+                   (throw (ex-info (tru "Query cancelled")
+                                   {:sql sql :parameters parameters ::cancelled? true}))
+                   BigQueryException
+                   (let [bqe ^BigQueryException value]
+                     (if (.isRetryable bqe)
+                       (throw (ex-info (tru "BigQueryException executing query")
+                                       {:retryable? (.isRetryable bqe)
+                                        :sql        sql
+                                        :parameters parameters}
+                                       bqe))
+                       (throw-invalid-query bqe sql parameters)))
+                   Throwable
+                   (throw-invalid-query value sql parameters)))
+
+        ;; cancel-chan fired first, it either got closed (nil) or received a real cancel message.
+        (if res
+          ;; Real cancel
+          (do (log/trace "BigQuery cancellation channel fired - cancelling the future if it's not already done")
+              (reset! cancel-requested? true)
+              (when-not (or (future-cancelled? res-fut) (future-done? res-fut))
+                (log/trace "BigQuery initial call still running, cancelling the future")
+                (future-cancel res-fut))
```

Comment:
I guess we're not canceling jobs upstream, right? Apparently there's a way to do it. https://cloud.google.com/bigquery/docs/samples/bigquery-cancel-job

Not sure since the sample code isn't in Java but it might be as simple as `(.cancel job)` in Java (assuming you can record the `Job` you submit somehow). If it's that easy we should do it while we're at it. If not then we should open a GH issue for this. LMK what you think